File: leetcode/editor/cn/421_maximum-xor-of-two-numbers-in-an-array.cpp
//Given an integer array nums, return the maximum result of nums[i] XOR nums[j],
// where 0 <= i ≤ j < n. 
//
// Follow up: Could you do this in O(n) runtime? 
//
// 
// Example 1: 
//
// 
//Input: nums = [3,10,5,25,2,8]
//Output: 28
//Explanation: The maximum result is 5 XOR 25 = 28. 
//
// Example 2: 
//
// 
//Input: nums = [0]
//Output: 0
// 
//
// Example 3: 
//
// 
//Input: nums = [2,4]
//Output: 6
// 
//
// Example 4: 
//
// 
//Input: nums = [8,10,2]
//Output: 10
// 
//
// Example 5: 
//
// 
//Input: nums = [14,70,53,83,49,91,36,80,92,51,66,70]
//Output: 127
// 
//
// 
// Constraints: 
//
// 
// 1 <= nums.length <= 2 * 104 
// 0 <= nums[i] <= 231 - 1 
// 
// Related Topics 位运算 字典树 
// 👍 359 👎 0


#include <cstdio>
#include <vector>
#include <algorithm>
#include <iostream>
using namespace std;

//leetcode submit region begin(Prohibit modification and deletion)
struct Trie {
    // 左子树指向表示 0 的子节点
    Trie* left = nullptr;
    // 右子树指向表示 1 的子节点
    Trie* right = nullptr;

    Trie() = default;
};
class Solution {
protected:
    Trie* root = new Trie();
    static constexpr int HIGH_BIT = 30;
public:
    void add(int num){
        Trie* cur = root;
        for (int k = HIGH_BIT; k >= 0; --k) {
            int bit = (num >> k) & 1;
            if(bit==0){
                if(!cur->left){
                    cur->left = new Trie();
                }
                cur = cur->left;
            }else{
                if(!cur->right){
                    cur->right = new Trie();
                }
                cur = cur->right;
            }
        }
    }

    int find(int num){
        Trie* cur = root;
        int x = 0;
        for (int k = HIGH_BIT; k >= 0; --k) {
            int bit = (num >> k) & 1;//高位到低位
            if (bit == 0) {
                //优先往right走
                if (cur->right) {
                    cur = cur->right;
                    x = x * 2 + 1;
                }
                else {
                    cur = cur->left;
                    x = x * 2;
                }
            }
            else {
                //优先往left走
                if (cur->left) {
                    cur = cur->left;
                    x = x * 2 + 1;
                }
                else {
                    cur = cur->right;
                    x = x * 2;
                }
            }
        }
        return x;
    }
    int findMaximumXOR(vector<int>& nums) {
        int n= nums.size();
        int x = 0;
        for(int i=1;i<n;i++){
            add(nums[i-1]);
            x = max(x, find(nums[i]));
        }
        return x;
    }
};
//leetcode submit region end(Prohibit modification and deletion)

File: leetcode/editor/cn/421_maximum-xor-of-two-numbers-in-an-array.php
<?php
//给你一个整数数组 nums ，返回 nums[i] XOR nums[j] 的最大运算结果，其中 0 ≤ i ≤ j < n 。 
//
// 进阶：你可以在 O(n) 的时间解决这个问题吗？ 
//
// 
//
// 
// 
// 示例 1： 
//
// 
//输入：nums = [3,10,5,25,2,8]
//输出：28
//解释：最大运算结果是 5 XOR 25 = 28. 
//
// 示例 2： 
//
// 
//输入：nums = [0]
//输出：0
// 
//
// 示例 3： 
//
// 
//输入：nums = [2,4]
//输出：6
// 
//
// 示例 4： 
//
// 
//输入：nums = [8,10,2]
//输出：10
// 
//
// 示例 5： 
//
// 
//输入：nums = [14,70,53,83,49,91,36,80,92,51,66,70]
//输出：127
// 
//
// 
//
// 提示： 
//
// 
// 1 <= nums.length <= 2 * 104 
// 0 <= nums[i] <= 231 - 1 
// 
// 
// 
// Related Topics 位运算 字典树 
// 👍 359 👎 0


//leetcode submit region begin(Prohibit modification and deletion)
class Trie{
    public $left = null;
    public $right  =null;
}
class Solution {
    public $trie;
    function add($num){
        $high = 30;
        $cur = $this->trie;
        for($k=$high;$k>=0;$k--){
            $v = ($num>>$k)&1;
            if($v==1){
                if(!$cur->right){
                    $cur->right = new Trie();
                }
                $cur = $cur->right;
            }else{
                if(!$cur->left){
                    $cur->left = new Trie();
                }
                $cur = $cur->left;
            }
        }
    }

    function find($num){
        $high = 30;
        $cur = $this->trie;
        $x = 0;
        for($k=$high;$k>=0;$k--){
            $v = ($num>>$k)&1;
            if($v==1){
                if($cur->left){
                    $cur = $cur->left;
                    $x = $x*2+1;
                }else{
                    $cur = $cur->right;
                    $x = $x*2;
                }
            }else{
                if($cur->right){
                    $x = $x*2+1;
                    $cur = $cur->right;
                }else{
                    $cur = $cur->left;
                    $x = $x*2;
                }
            }
        }
        return $x;
    }

    /**
     * @param Integer[] $nums
     * @return Integer
     */
    function findMaximumXOR($nums) {
        $this->trie = new Trie();
        $x = 0;
        foreach ($nums as $num){
            $this->add($num);
            //print_r($this->trie);
            $x = max($x,$this->find($num));
        }
        return $x;
    }
}

//$s = new Solution();
//echo $s->findMaximumXOR([3,10,5,25,2,8]).PHP_EOL;

//leetcode submit region end(Prohibit modification and deletion)
